### PR TITLE
Seasons

### PR DESCRIPTION
**Pull request by QSparks**
This pull request is for an extension of the climdexInput data structure to allow the calculation of climate indices at a seasonal level.

Overview:
Extended the climdexInput data structure to support meteorological seasons as a recognized frequency for calculating climate indices. This definition of seasons includes handling for the winter season, where data in the months of January and February are assigned to the winter season of the previous year. The new seasonal aggregation frequency is applied to the time-series data for various climate variables, including `tmax`, `tmin`, `prec`, and `tavg`.

NA masks are generated for seasonal data so that any combination of season and climate variables missing more than 6 days of climate data will be classified as an NA season-variable pair and will not be available for calculating climate indices.

Indices with new seasonal aggregation availability include: `tnn, tnx, txn, txx, dtr, tn10p, tn90p, tx10p, tx90p`

Scope of Changes:
Updated functions related to seasonal data aggregation in `climdex.R`.
Added testing in `test_seasonal_date_factor.R` for the new seasonal aggregation. 

Testing:
Testing was performed on NA season assignment and seasonal index calculations.
All tests passed successfully using the sample datasets.

Documentation:
Updated relevant sections of the project documentation to include the seasonal date factor.

Reviewer notes: 
Please review the logic for marking NA seasons. 
I would appreciate feedback on the decision of `max.missing.days = 6`

### CODE REVIEW

File: R/climdex.r [hunk 1]
```diff
@@ -560,10 +563,31 @@ climdexInput.raw <- function(tmax=NULL, tmin=NULL, prec=NULL, tmax.dates=NULL, t
   new.date.range <- as.PCICt(paste(as.numeric(format(range(all.dates), "%Y", tz="GMT")), c("01-01", last.day.of.year), sep="-"), cal=cal)
   date.series <- seq(new.date.range[1], new.date.range[2], by="day")
   jdays <- get.jdays.replaced.feb29(get.jdays(date.series))
-  
-  ## Factors for dividing data up
-  date.factors <- list(annual=factor(format(date.series, format="%Y", tz="GMT")), monthly=factor(format(date.series, format="%Y-%m", tz="GMT")))
 
+  # Classify meteorological seasons with associated years
```
Comment (corviday) [inline R/climdex.r:567]: I see you have a lot of special handling for winter and the year boundary, that's good. 

What happens to the first year in the dataset, does it have  winter measurements taking only two months into account? Is this documented anywhere?

File: R/climdex.r [hunk 2]
```diff
@@ -764,7 +802,7 @@ climdex.id <- function(ci) { stopifnot(!is.null(ci@data$tmax)); return(number.da
 #' daily minimum temperature stays above 20 degrees Celsius.
 #' 
 #' @param ci Object of type climdexInput.
-#' @return A vector containing the number of frost days for each year.
+#' @return A vector containing the number of tropical nights for each year.
```
Comment (corviday) [inline R/climdex.r:805]: nice catch!

File: tests/test_seasonal_date_factor.R [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+datacol <- list(tmax = "MAX_TEMP", tmin = "MIN_TEMP", prec = "ONE_DAY_PRECIPITATION")
+ci.csv <- climdexInput.csv(ClimVars.tmax, ClimVars.tmin, ClimVars.prec, data.columns = datacol, base.range = c(1981, 1990))
+
+unique_seasons <- unique(ci.csv@date.factors$seasonal)
+unique_months <- unique(ci.csv@date.factors$monthly)
+
+climdex.pcic.test.seasonal.na.cases <- function() {
+  var.list <- c("tmax", "tmin", "prec", "tavg")
+  for (var in var.list){
+    # Test 1.1: Check if the first season is NA based on the start of the time series
```
Comment (corviday) [inline tests/test_seasonal_date_factor.R:20]: Ah, this answers my earlier question. Good thing to have a test for.

**Review by corviday — COMMENTED** [review]
Code looks good, I appreciate how much work went into updating all the documentation, and adding tests is always nice!

I wanted to run the tests with R CMD check, but theydon't seem to be set up to be run that way. How are you running your tests. Do you plan to update them at some point in the future to be run automatically by R's fussy systems?

**Comment by QSparks** [comment]
> Code looks good, I appreciate how much work went into updating all the documentation, and adding tests is always nice!
> 
> I have a couple of questions about fussy R package formatting things, but they might be outside the scope of this PR, or maybe the ultimate plan isn't to resubmit this package to CRAN:
> 
> 1. I wanted to run the tests with `R CMD check`, but they don't seem to be set up to be run that way. How are you running your tests? Do you plan to update them at some point in the future to be run automatically by R's fussy systems?
> 2. Do you need to update the DESCRIPTION or other R-package related things?

Thanks, Lee.
I haven’t run into issues running
 `R CMD check climdex.pcic` or 
`R CMD build climdex.pcic` followed by `R CMD check climdex.pcic_1.1-11.tar.gz`
from the root directory of the package. 

I have also been able to run the test suite using RStudio's build tools.

You are right, I will need to update the package `DESCRIPTION` and `CHANGLOG`. There are a few updates planned so it might be worth promoting the package to version 2. For now, I will add a version increment for each additional feature.

File: R/climdex.r (hunk 1 above)
Comment (QSparks) [reply R/climdex.r:567]: I will add a clarification to the documentation. Only seasons where all three constituent months are valid, and the sum of NA days for the season is no greater than six will be valid. This means that a time series starting in January and ending in December will have NA winter seasons at both ends of the series.

File: R/climdex.r [hunk 4]
```diff
@@ -537,14 +552,14 @@ get.outofbase.quantiles <- function(tmax=NULL, tmin=NULL, prec=NULL, tmax.dates=
 #' ec.1018935.tmin$MIN_TEMP, ec.1018935.prec$ONE_DAY_PRECIPITATION,
 #' tmax.dates, tmin.dates, prec.dates, base.range=c(1971, 2000))
 #'
-#' @export
+#' @export 
 climdexInput.raw <- function(tmax=NULL, tmin=NULL, prec=NULL, tmax.dates=NULL, tmin.dates=NULL, prec.dates=NULL,
                              base.range=c(1961, 1990), n=5, northern.hemisphere=TRUE,
-                             tavg=NULL, tavg.dates=NULL, quantiles=NULL, temp.qtiles=c(0.10, 0.90), prec.qtiles=c(0.95, 0.99), max.missing.days=c(annual=15, monthly=3), min.base.data.fraction.present=0.1) {
+                             tavg=NULL, tavg.dates=NULL, quantiles=NULL, temp.qtiles=c(0.10, 0.90), prec.qtiles=c(0.95, 0.99), max.missing.days=c(annual=15, monthly=3, seasonal=6), min.base.data.fraction.present=0.1) {
```
Comment (jameshiebert) [inline R/climdex.r:558]: 6 days per season sounds great.

File: R/climdex.r [hunk 5]
```diff
@@ -560,10 +575,39 @@ climdexInput.raw <- function(tmax=NULL, tmin=NULL, prec=NULL, tmax.dates=NULL, t
...
+  # Classify meteorological seasons with associated years. This includes 
+  # handling for the winter season, where data in the months of January and 
+  # February are assigned to the winter season of the previous year.
+  # 
+  # The season description is as follows:
+  #  - 'Winter': December, January, February
+  #  - 'Spring': March, April, May
+  #  - 'Summer': June, July, August
+  #  - 'Fall': September, October, November
+  classify_meteorological_season_with_year <- function(date.series) {
```
Comment (jameshiebert) [inline R/climdex.r:588]: Have we yet introduced you to climate model's "idealized calendars"? Some calendars ("365-day") will omit all Feb 29s in leap years, and some calendars only contain 360 days per year. In the latter instance, if this function receives a properly attributed PCICt object with a 360 calendar, how are the seasons mapped and is it reasonable?

File: R/climdex.r [hunk 6]
```diff
@@ -958,14 +1016,14 @@ climdex.tn10p <- function(ci, freq=c("monthly", "annual")) { stopifnot(!is.null(
 #' @template get_generic_example
 #' 
 #' @export
-climdex.tx10p <- function(ci, freq=c("monthly", "annual")) { stopifnot(!is.null(ci@data$tmax) && !is.null(ci@quantiles$tmax)); return(percent.days.op.threshold(ci@data$tmax, ci@dates, ci@jdays, ci@date.factors[[match.arg(freq)]], ci@quantiles$tmax$outbase$q10, ci@quantiles$tmax$inbase$q10, ci@base.range, "<", ci@max.missing.days[match.arg(freq)]) * ci@namasks[[match.arg(freq)]]$tmax) }
+climdex.tx10p <- function(ci, freq=c("monthly", "annual", "seasonal")) { stopifnot(!is.null(ci@data$tmax) && !is.null(ci@quantiles$tmax)); return(percent.days.op.threshold(ci@data$tmax, ci@dates, ci@jdays, ci@date.factors[[match.arg(freq)]], ci@quantiles$tmax$outbase$q10, ci@quantiles$tmax$inbase$q10, ci@base.range, "<", ci@max.missing.days[match.arg(freq)]) * ci@namasks[[match.arg(freq)]]$tmax) }
```
Comment (jameshiebert) [inline R/climdex.r:1019]: All of the changes are pretty straightforward once you get the factors computed, eh? That's a very good indication that you solved the problem in a good way :)

File: tests/test_seasonal_date_factor.R [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    # - If the first month of the date series is not a multiple of 3, expect the first season to be NA
+    checkTrue((as.integer(format(ci.csv@dates[1], "%m")) %% 3 != 0) && is.na(ci.csv@namasks$seasonal[[var]][1]))
```
Comment (jameshiebert) [inline tests/test_seasonal_date_factor.R:22]: There's a lot going on in these checks. Any way that you can break this up a bit more for readability?

File: tests/test_seasonal_date_factor.R [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+
+climdex.pcic.test.seasonal.indices <- function() {
+  # Test 2:
+  # This section handles various scenarios based on the type of index being calculated:
+  # - For non-averaged stats (e.g., min, max), the seasonal value is the min/max of the constituent months.
+  # - For averaged stats, the seasonal value may differ from the average of the 3 constituent months due to varying month lengths, so it's compared against a calculated dtr (daily temperature range).
+  # - For percentile calculations, each season's percentile is computed against a weighted monthly percentile.
+
+  min_max <- list(
```
Comment (jameshiebert) [inline tests/test_seasonal_date_factor.R:56]: These categories that you've put together are pretty useful and seems like they would have broader utility than only in the tests. What do you think about including them as package-level constants?

File: tests/test_seasonal_date_factor.R [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  ci_subset <- function(ci.csv, dates, season) {
+    return(climdexInput.raw(
+      tmax = ci.csv@data$tmax[ci.csv@date.factors$seasonal == season],
+      tmin = ci.csv@data$tmin[ci.csv@date.factors$seasonal == season],
+      prec = ci.csv@data$prec[ci.csv@date.factors$seasonal == season],
+      tavg = ci.csv@data$tmin[ci.csv@date.factors$seasonal == season],
+      tmax.dates = dates, tmin.dates = dates, prec.dates = dates, tavg.dates = dates, base.range = c(1981, 1990)
+    ))
+  }
+  for (index in seasonal_indices) {
```
Comment (jameshiebert) [inline tests/test_seasonal_date_factor.R:73]: I'm not keen on the structure of this loop. While it covers a lot of different cases, there is too much nested logic and switches to be able to easily follow the program flow. Can you break this up in a different way?

**Review by jameshiebert — COMMENTED** [review]
Looks really good @QSparks ! I think that you found the right structure to make your changes, as you pretty much had to add a couple functions to create the date factors and everything else (testing aside) was pretty trivial.

I think the testing could use a bit more work for clarity, but overall it looks great!

File: tests/test_seasonal_date_factor.R (hunk 9 above)
Comment (QSparks) [reply tests/test_seasonal_date_factor.R:73]: Yes, I will break up the nested logic, and run separate tests for each type of index (min-max/mean/bootstrap).

File: tests/test_seasonal_date_factor.R (hunk 8 above)
Comment (QSparks) [reply tests/test_seasonal_date_factor.R:56]: Agreed, I will add new constants.

File: R/climdex.r (hunk 6 above)
Comment (QSparks) [reply R/climdex.r:1019]: Thanks, James!

**Comment by QSparks** [comment]
I believe all feedback for this PR has been addressed and, once approved, can be closed as it is included in https://github.com/pacificclimate/climdex.pcic/pull/32. Note that additional feature branches have since been created including a refactoring step in PR, including https://github.com/pacificclimate/climdex.pcic/pull/36 that breaks climdex.r apart into more manageable R scripts. This will make rebasing a bit nasty, and I propose any additional change requests be documented as new issues for correction off the modular version that is under review. Those changes will be addressed promptly as active development has resumed for this project.

**Review by corviday — APPROVED** [review]
Agree that additional changes at this point will only cause thrice the amount of trouble in merging. Looks good to me.

**Pull request merged**